ophyd/sim.py

```diff
@@ -226,30 +226,78 @@ def __init__(self, func=None, *,
                  **kwargs):
         if func is None:
             func = np.random.rand
+
+        self._period = period
+        self._period_jitter = period_jitter
+
         super().__init__(name=name, func=func,
                          exposure_time=exposure_time,
                          parent=parent, labels=labels, kind=kind,
                          **kwargs)
+        self.__thread = None
+
+    def start_simulation(self):
+        """
+        Start background thread that performs periodic PV updates. The method
+        should be called at least once before the beginning of simulation. Multiple
+        calls to the method are ignored.
+        """
+        if self.__thread is None:
+
+            def periodic_update(ref, period, period_jitter):
+                while True:
+                    signal = ref()
+                    if not signal:
+                        # Our target Signal has been garbage collected. Shut
+                        # down the Thread.
+                        return
+                    signal.put(signal._func())
+                    del signal
+                    # Sleep for period +/- period_jitter.
+                    ttime.sleep(
+                        max(self._period + self._period_jitter * np.random.randn(), 0))
+
+            self.__thread = threading.Thread(target=periodic_update,
+                                             daemon=True,
+                                             args=(weakref.ref(self),
+                                                   self._period,
+                                                   self._period_jitter))
+            self.__thread.start()
+
+    def _start_simulation_deprecated(self):
+        """Call `start_simulation` and print deprecation warning"""
+        if self.__thread is None:
+            msg = ("Deprecated API: Objects of SynPeriodicSignal must be initialized before simulation\n"
+                   "by calling 'start_simulation()' method. Two such objects ('rand' and 'rand2') are\n"
+                   "created by 'ophyd.sim' module. Call\n"
+                   "    rand.start_simulation() or rand2.start_simulation()\n"
+                   "before the object is used.")
+            logger.warning(msg)
```

Comment:
Nit pick: I think using `self.log`, defines by the base class `OphydObj`, can be used here to provide additional info—namely, which instance this warning is being reported from.

Reply:
Yes, I think it's a good idea to replace the logger. The only thing is that I thought that the deprecation warning should be displayed to the user no matter what and logger must be explicitly activated, which may not be done for simulated runs (examples). Or the logs may go into a file. If we want users to update the examples, then may be it's a good idea to print the message as well?

Reply:
>  logger must be explicitly activated

Python's default configuration includes [a "handler of last resort"](https://docs.python.org/3/library/logging.html#logging.lastResort) which writes messages of level WARNING or higher to the stderr. Thus, the warnings here _will_ be shown unless the message is processed by an explicitly registered handler that directs it elsewhere. In the case our tutorials/examples, there are no other handlers registered. In the case of beamlines, I believe that WARNING and higher are shown in the stderr and INFO and higher are logged to a file.

We generally avoid printing because there is no way for downstream libraries to silence prints.